keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,705 @@
+expected/allowed to contain IPv6 addresses in a single-stack IPv4
+cluster, or IPv4 addresses in a single-stack IPv6 cluster. For cloud
+providers that do not currently return IPv6 addresses, starting to
+return them might break clients which were written without taking this
+possibility into account.
+
+[the Kubernetes network model]: https://kubernetes.io/docs/concepts/cluster-administration/networking/#the-kubernetes-network-model
+
+#### Address Detection Behavior Across Different Cluster Types
+
+On bare-metal, a node always has exactly one IP (even if the cluster
+is supposed to be dual-stack), which is auto-detected by kubelet, but
+can be overridden with `--node-ip`.
+
+When using a legacy cloud provider, the node IPs are detected by the
+cloud provider, and it may return multiple IPs of each type (eg, if
+the node has multiple network interfaces, or multiple addresses on a
+single interface). eg:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.1.2.3" },
+      { "type": "InternalIP", "address": "10.4.5.6" },
+      { "type": "InternalIP", "address": "10.7.8.9" } ]
+
+In this case, the primary IP would be `10.1.2.3`. If the administrator
+wants a different primary IP, they can pass `--node-ip` to choose
+another one, but that will have the side effect of removing all other
+IPs of the same type from the list. Eg, passing `--node-ip 10.4.5.6`,
+would result in `Node.Status.Addresses` having the value:
+
+    [ { "type": "ExternalIP", "address": "42.42.42.42" },
+      { "type": "InternalIP", "address": "10.4.5.6" } ]
+
+When using an external cloud-provider, the administrator has no
+control over the node IPs; in the above example, passing `--node-ip
+10.4.5.6` would have no effect on the returned list of addresses, and
+the primary IP would still be `10.1.2.3`. (Although passing `--node-ip
+10.11.12.13` would result in an error since that doesn't appear in the
+list at all.)
+
+#### API Cleanliness
+
+`Node.Status.Addresses` is annotated with an incorrect strategic-merge
+patch annotation, and so it is not possible to use strategic merge
+patch to change that field without losing information. (See discussion
+in [kubernetes #79391].) Although kubelet and cloud-controller-manager
+have been fixed to work around the problem, it is still annoying to
+have the incorrect annotation there, and it is possible that other
+components might attempt to update the field incorrectly in the
+future.
+
+Additionally, `Node.Status.Addresses` is not currently sufficiently
+checked by validation; it is possible for an external cloud provider
+to declare an `InternalIP` or `ExternalIP` address that is not a
+syntactically-valid IP address, meaning clients that read this field
+may need to sanity-check it themselves ([kubernetes #89817]). Adding
+validation now could break deployed cloud providers (which might
+otherwise appear to be fully-functional, as long as the primary IP was
+valid).
```

Comment:
So first off, it doesn't _always_ have to be parsed; sometimes you're passing it to another API that takes a string (eg, creating an iptables rule that uses the node's IP).

But if it does have to be parsed, and there is legitimately reason why the parse might fail, then you have to add in error handling to deal with that case (eg, if the code that needs to know the IP finds an invalid IP, then what should other parts of the code that don't care about the IP do? Should they also ignore the Node until the IP is fixed? etc). OTOH, if the field is checked by validation, then parse failure "can't happen" and you don't need to worry about responding in a consistent way if it does (because presumably everything else is also broken and the whole thing is about to fall over anyway).